src/util/segmentation/operations/correction.js

```diff
@@ -0,0 +1,524 @@
+ * @param  {Object} evt The cornerstone event.
+ * @param  {number} segmentIndex
+ * @returns {boolean} Returns true if the operation was simple.
+ */
+function simpleScissorOperation(nodes, evt, segmentIndex) {
+  let allInside = true;
+  let allOutside = true;
+
+  for (let i = 0; i < nodes.length; i++) {
+    const node = nodes[i];
+
+    if (node.segment === segmentIndex) {
+      allOutside = false;
+    } else {
+      allInside = false;
+    }
+
+    if (!allInside && !allOutside) {
+      break;
+    }
+  }
+
+  if (allOutside) {
+    logger.warn('The line never intersects a segment.');
+    fillInside(evt);
+
+    return true;
+  } else if (allInside) {
+    logger.warn('The line is only ever inside the segment.');
+    eraseInside(evt);
+
+    return true;
+  }
+
+  return false;
+}
+
+/**
+ * performOperation - Performs the given add/subtract operation using a modification of the Tobias Heimann Correction Algorithm:
+ * The algorithm is described in full length in Tobias Heimann's diploma thesis (MBI Technical Report 145, p. 37 - 40).
+ *
+ * @param  {Object} operation The operation.
+ * @param  {UInt16Array} pixelData The 2D labelmap.
+ * @param  {UInt16Array} workingLabelMap A copy of the labelmap for processing purposes.
+ * @param  {number} segmentIndex The label of the tool being used.
+ * @param  {Object} evt The cornerstone event.
+ */
+function performOperation(
+  operation,
+  pixelData,
+  workingLabelMap,
+  segmentIndex,
+  evt
+) {
+  const eventData = evt.detail;
+  const { image } = eventData;
+  const cols = image.width;
+  const rows = image.height;
+
+  const { nodes, additive } = operation;
+  const fillOver = additive ? 0 : 1;
+
+  // Local getters to swap from cornerstone vector notation and flattened array indicies.
+  const getPixelIndex = pixelCoord => pixelCoord.y * cols + pixelCoord.x;
+  const getPixelCoordinateFromPixelIndex = pixelIndex => ({
+    x: pixelIndex % cols,
+    y: Math.floor(pixelIndex / cols),
+  });
+
+  if (additive) {
+    logger.warn('additive operation...');
+  } else {
+    logger.warn('subtractive operation...');
+  }
+
+  const { pixelPath, leftPath, rightPath } = getPixelPaths(nodes);
+
+  // Find extent of region for floodfill (This segment + the drawn loop).
+  // This is to reduce the extent of the outwards floodfill, which constitutes 99% of the computation.
+  const firstPixelOnPath = pixelPath[0];
+
+  const boundingBox = {
+    xMin: firstPixelOnPath.x,
+    xMax: firstPixelOnPath.x,
+    yMin: firstPixelOnPath.y,
+    yMax: firstPixelOnPath.y,
+  };
+
+  // ...whilst also initializing the workingLabelmap
+  for (let i = 0; i < workingLabelMap.length; i++) {
+    if (pixelData[i] === segmentIndex) {
+      const pixel = getPixelCoordinateFromPixelIndex(i);
+
+      expandBoundingBox(boundingBox, pixel);
+      workingLabelMap[i] = 1;
+    } else {
+      workingLabelMap[i] = 0;
+    }
+  }
+
+  // Set workingLabelmap pixelPath to 2 to form a boundary.
```

Comment:
I'm not sure I follow this comment.

Reply:
I'll clarify in-code. Its set as a different label in the workingLabelmap so that we can perform floodfills either side of it. We store it as 2 (rather than 0) so that we can fill it with the appropriate color later.

Reply:
2 is special because it's generally unused?

Reply:
The working label map is just a temporary object, 2 is used just because I want to keep track of a different entity (the path), and 0 and 1 are used.